### ISSUE TITLE
no-param-reassign should not throw error when using Object destructuring on parameters

### ISSUE THREAD

**Issue by kaiye** [eslint/eslint#11088]
**Tell us about your environment**

* **ESLint Version:** 5.9.0
* **Node Version:** v8.11.3
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "parserOptions": {
    "ecmaVersion": 2018,
    "sourceType": "module"
  },
  "rules": {
    "no-param-reassign": [
      "error",
      {
        "props": true
      }
    ]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const obj = { key: [1, 2, 3] };

function foo({ key }) {
  key = [4, 5, 6];
  return key;
}

function bar(qux) {
  let { key } = qux;
  key = [4, 5, 6];
  return key;
}
```

```bash
node_modules/eslint/bin/eslint.js demo.js
```

**What did you expect to happen?**
0 error

**What actually happened? Please include the actual, raw output from ESLint.**
`4:3  error  Assignment to function parameter 'key'  no-param-reassign`

**Comment by kaiye** [comment]
The `foo` function has the same meaning as the `bar` function, there do not reassign the argument passed in.
However, `foo` throws error, but `bar` doesn't.

**Comment by nzakas** [comment]
Thanks for the report. I believe that this is the way the rule is intended to work based on this test:
https://github.com/eslint/eslint/blob/master/tests/lib/rules/no-param-reassign.js#L66

**Comment by pantoninho** [comment]
I'm experiencing the same problem. Why should this lint rule throw on this situation?

I might be wrong but when we use object destructuring on the arguments, we are creating mutable local variables that point to the value of the object property of the same name.

I've read that this rule prevents bugs related to how javascript handles the arguments and the argument object of the function (explained here: https://spin.atomicobject.com/2011/04/10/javascript-don-t-reassign-your-function-arguments/), but I think that re-assignment of destructured properties of an argument falls outside the scope of the problem mentioned in the article.


On a side note, I think this rule should only prevent mutations on arguments (that are objects/arrays). The problem described in that article (that is linked on this rule's doc page) could be entirely avoided by preventing the use of the `arguments` object -- which is completely possible nowadays and would not throw on perfectly valid re-assignments like:
```
function withMinValue(x) {
   x = x > 0.05 ? x : 0.05;
   // do things with x
}
``` 
or
```
function withDefaultValue(x) {
 x = x.value || x || getDefaultValue();
 // do things with x;
}
```

**Comment by platinumazure** [comment]
@pantoninho Your premise about modifying the destructured parameter is false, per [this demo](https://jsfiddle.net/hv5kpos1/).

Please feel free to open a new issue regarding your enhancement request about only modifying objects/arrays. (Although, that said, I think we'll probably say we can't implement it in core ESLint due to lack of type knowledge.)

**Comment by pantoninho** [comment]
Hello @platinumazure. Thanks for the quick reply. 

Maybe I did not explain myself correctly, but that demo does not make my premise false. 

Your function ‘modifyDestructure’ creates a local variable ‘x’ which points to the value ‘“foo”’. Changing the value of that local variable has no effect outside of that function scope and (I need to confirm this one) on the arguments object — which is actually the cause of the eslint rule as per the docs. 

Going back to the original question, why would your ‘modifyDestructure’ function throw a lint error?

**Comment by platinumazure** [comment]
@pantoninho Possible misreading on my part:

> I might be wrong but when we use object destructuring on the arguments, we are creating mutable local variables that point to the value of the object property of the same name.

Guess I took that out of context and read that as an assertion about the leakage of assignments to an object property "through" an assignment to a destructured value. My mistake.

In any case, I still recommend creating a new issue, so that you can provide full context for your use case.

**Comment by pantoninho** [comment]
@platinumazure I can understand why it was easy to misread, sorry.

I will follow your advice.
Thanks
